src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -851,20 +851,32 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
             return http2Stream;
         }
 
-        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer)
+        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
         {
             ReadOnlyMemory<byte> remaining = buffer;
 
             while (remaining.Length > 0)
             {
                 int frameSize = Math.Min(remaining.Length, FrameHeader.MaxLength);
 
-                frameSize = await _connectionWindow.RequestCreditAsync(frameSize).ConfigureAwait(false);
+                // Once credit had been granted, we want to actually consume those bytes.
+                frameSize = await _connectionWindow.RequestCreditAsync(frameSize, cancellationToken).ConfigureAwait(false);
 
                 ReadOnlyMemory<byte> current;
                 (current, remaining) = SplitBuffer(remaining, frameSize);
 
-                await AcquireWriteLockAsync().ConfigureAwait(false);
+                // It's possible that a cancellation will occur while we wait for the write lock. In that case, we need to
+                // return the credit that we have acquired and don't plan to use.
+                try
+                {
+                    await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
```

Comment:
Is this specific to cancellation?  Or would we do that for any exception?

Reply:
Any exception here other than cancellation should cause the connection to be killed, so I don't think it matters.

That said, I think this is probably related to the question of how we handle cancellation in FlushOutgoingBytesAsync....

Reply:
I should add, it's a little odd that we have the try/finally below where we call ReleaseWriterLock in the finally. If we throw in that block, the connection is going to be killed, as mentioned above. But we still need to ensure writers are unblocked, even though they are going to fail their writes.

An alternative here would be to just dispose the writer lock in FlushOutgoingBytesAsync if a write exception occurs. That would cause any waiting writers to throw, which is probably slightly better than having them unblock, try to write, and then fail. And we wouldn't need the try/finally here.

Reply:
I think that the specific try-catch referenced here is necessary. Cancellation at this point is not necessarily going to kill the connection, so we need to release the credit acquired by this request. Otherwise we might end up hanging if our credit accounting differs from the servers.

I don't have an issue removing the try-finally block below. It does mean that any exceptions during request serialization will be fatal to the connection, but that seems like a reasonable behavior.